content/post/2019-05-18-talking-i2c.md

```diff
@@ -0,0 +1,145 @@
+
+```elixir
+{:ok, bus} = Circuits.I2C.open("i2c-1")
+:ok = Circuits.I2C.write(bus, 72, <<1>>)
+{:ok, <<133, 131>>} = Circuits.I2C.read(bus, 72, 2)
+# I2C also has a convenience method for this
+{:ok, <<133, 131>>} = Circuits.I2C.write_read(bus, 72, <<1>>, 2)
+```
+
+We're getting back some data, but what do those bytes mean?
+We'll need to take a look at [the datasheet for the ADS1115](http://www.ti.com/lit/ds/symlink/ads1115.pdf) chip.
+These 16 bits of data tell us 9 different things about how the chip is currently operating.
+That's a lot of information packed into the space of just two ascii characters.
+
+![Config Register](/images/2019-05-18/ADS1115_Config_Register.png)
+
+Luckily, we can use [Elixir's binary pattern matching syntax](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#%3C%3C%3E%3E/1) to pull out this information.
+Binary pattern matching allows us to ask for information by expressing how we expect the information to be laid out across the bytes.
+No bit-shifting required.
+
+```elixir
+<< status::size(1), mux::size(3), _other_stuff::size(12) >> = <<133, 131>>
+```
+
+We can also similarly build a new binary to write new configuration to the register.
+
+```elixir
+config_bytes = << status::size(1), mux::size(3), other_stuff::size(12) >>
+Circuits.I2C.write(bus, 72, <<1>> <> config_bytes)
+```
+
+## Sharing What We've Learned
+
+The binary pattern matching syntax is really helpful for parsing binary data, but learning which bits do which things still requires some squinting and a datasheet.
+So I went ahead and published the [ADS1115 package on hex.pm](https://hexdocs.pm/ads1115/0.1.0/readme.html) so that other people can deal with a struct of config data.
+
+```elixir
+{:ok, bus} = Circuits.I2C.open("i2c-1")
+ADS1115.config(bus, 72)
+=> {:ok, %ADS1115.Config{
+  performing_conversion: false,
+  mux: {:ain0, :ain1},
+  gain: 2048,
+  mode: :single_shot,
+  data_rate: 128,
+  comp_mode: :traditional,
+  comp_polarity: :active_low,
+  comp_latch: false,
+  comp_queue: :disabled
+}}
+```
+
+The package also gives you an easy way to read the sensor.
+Now I'm ready to check my soil moisture sensors.
+
+## Will This Plant Live?
+
+Technology is no guarantee of success, but this is the best chance I have at keeping a plant alive.
+I'll be doing some basic experiments around how much moisture different plants need and whether they like to have dry cycles, but here's a simplistic example of to monitor and water a plant.
+
+```elixir
+defmodule Seedling do
+  use GenServer
+
+  @a2d_addr 72 # the address of our ADS1115 analog to digital converter
+  @check_interval 1_000 # wait 1000ms between checking the sensor
+  @pump_control_pin 4 # the GPIO pin we are using to control our water pump
+  @sensor_threshold 23_989 # a little experimenting showed that I get this reading or lower when soil is wet
+
+  def init do
+    {:ok, bus} = Circuits.I2C.open("i2c-1")
+    {:ok, pump} = Circuits.GPIO.open(@pump_control_pin, :output)
```

Comment:
Maybe add `initial_value: 0` to have the pump off so that it's in a known state?

Reply:
:+1: